### DISCUSSION TITLE
How can I add a new planning algorithm?

### DISCUSSION THREAD

**Discussion by JmingZhang** [frePPLe/frepple#380] [Q&A]
Dear Johan and other friends:
Recently I am trying to use frepple, but the planing algorithm could not meet our requirements. So, I was wandering how can I added a new algorithm or do some modification. Is there any document for guide? Or can some one tell me where is the function of the algorihtm?
Thanks

**Comment by jdetaeye** [comment]
Some notes on this topic:

- There are 2 main ways to plug in a new planning algorithm.
- The first method:  If you only need minor changes to the algorithm, you can leverage the existing solver and adjust where needed. This requires C++ coding and/or some python scripting.
- The second method: if your solver is very different, you wouldn't leverage the existing solver. Instead, your build your own executable that will read input data from the frepple dabase and export back the results.
- For both methods there is no documentation available. And we have no plans to improve that any time soon. You're invited to contribute to the documentation :-)
- For completeness: frepple is licensed under the AGPL license. That license means in short that "what is open source must remain open source".   You are not allowed closed source planning algorithms, but you need to keep your planning algorithm also open source.

**Reply by JmingZhang** [reply]
Thank you ! If I can figure it out and get it done, I will contribute it back.
